### DISCUSSION TITLE
Manually trigger `loader` refresh

### DISCUSSION THREAD

**Discussion by chadoh** [remix-run/react-router#10381] [Q&A] [answered]
I'm building an offline-first app where data is primarily stored in local storage using [localForage](https://github.com/localForage/localForage). Here's a simplified version of [my index route](https://github.com/chadoh/entire-life-offline-first/blob/feat/g-sheets-sync/src/react/routes/ledgers/root.tsx):

```tsx
import { useLoaderData } from 'react-router-dom'
import type { LoaderFunction } from '@remix-run/router'
import { getLedgers } from '../../../data/local'

export const loader: LoaderFunction = async (): Promise<string[]> => {
  return getLedgers()
}

function Root() {
  const ledgers = useLoaderData() as Awaited<string[]>
  return <>{ledgers.join(', ')}</>
}

export const element = <Root />
```

If the user signs into Google, a worker is added to the page, syncing data to/from Google Sheets in the background. As the worker updates the local data store, it sends an event to the main process. How can I manually/procedurally trigger a reload of the `loader` when this happens?

This also seems relevant to apps that use subscriptions with GraphQL, WebSockets, or other architectures.

**Comment by kiliman** [comment]
```ts
const navigate = useNavigate()
navigate('.', { replace: true })
```
https://reactrouter.com/en/main/hooks/use-navigate

**Reply by chadoh** [reply]
Wonderful, thank you.

I [created](https://github.com/chadoh/entire-life-offline-first/commit/536c720bd0e319601ef55f8beaa6968ed2f2a61b#diff-c795ff5da1b935bf480659412b8c47d433c358caab6b876358785fb771ae7a26R16-R23) a simple hook:

```ts
import { useNavigate } from 'react-router-dom'
import { subscribe } from '../data/local'

/**
 * Refresh data from this route's `loader` when data changes in the local data store.
 */
export function useDataSubscription() {
  const navigate = useNavigate()
  React.useEffect(() => {
    return subscribe(() => {
      navigate('.', { replace: true })
    })
  }, [])
}
```

Which can be used within a component. Diff with the component I showed above:

```diff
 import { useLoaderData } from 'react-router-dom'
 import type { LoaderFunction } from '@remix-run/router'
 import { getLedgers } from '../../../data/local'
+import { useDataSubscription } from '../../hooks'
 
 export const loader: LoaderFunction = async (): Promise<string[]> => {
   return getLedgers()
 }
 
 function Root() {
+  useDataSubscription()
   const ledgers = useLoaderData() as Awaited<string[]>
   return <>{ledgers.join(', ')}</>
 }
 
 export const element = <Root />
```

**Reply by kiliman** [reply]
Nice. @sergiodxa has a whole article on different techniques for refreshing data.

https://sergiodxa.com/articles/automatic-revalidation-in-remix

**Reply by sergiodxa** [reply, marked as answer]
Also you don't need to do the `navigate(".")` anymore, use the [useRevalidator hook](https://reactrouter.com/en/main/hooks/use-revalidator) instead

**Reply by chadoh** [reply]
@sergiodxa if you add that as a separate answer, I'll mark it as correct, so future readers don't use the less elegant API

**Reply by DanBotero** [reply]
@sergiodxa In Remix docs, the note for this hook says:

"While you can render multiple occurrences of useRevalidator at the same time, underneath it is a singleton. This means when one revalidator.revalidate() is called, all instances go into the "loading" state together (or rather, they all update to report the singleton state)."

Will the same behavior happen with the use of useNavigate()?

**Reply by sergiodxa** [reply]
useNavigate just returns a functions, and the navigation is global, so the state is reported in useNavigation which is also a singleton since it's the global navigation.

**Comment by nmitic** [comment]
I landed here after a google search for "retry loader manually for selected differed requests" 

Is something like this possible?

I have multiple request which are `deffered` and later `awaited` in the UI with loader and error component. But I would like to provide the users with the way to retry failed requests.

I know I can more fetching to the component, but I would like to know if I can do it using the react router loader instead and keep thing simpler.

**Reply by Clearmist** [reply]
I think I've figured out this problem. When you have deferred data fetched from a loader then want to refetch the data we can make use of `Promise.all` to properly show our loading component while the revalidation (rerunning our loader) is occurring.

```javascript
import { Suspense } from 'react';
import { Await, useAsyncValue, useLoaderData, useRevalidator } from 'react-router-dom';

function DisplayData () {
  const data = useAsyncValue();
}

export default function Leaderboard() {
  const revalidator = useRevalidator();
  const data = useLoaderData();

  const revalidatePromise = new Promise((resolve, reject) => {
    if (revalidator.state === 'idle') {
      resolve('now idle');
    }
  });

  // Create a Promise.all for the Await component to use. Include the revalidation promise, but keep only the results of the promise we want.
  const Promises = Promise.all([data.myData, revalidatePromise]).then((value) => value[0]);

  return (
    <div>
      <Suspense fallback={<Pending />}>
        <Await resolve={Promises} errorElement={<div>The component did not load correctly.</div>}>
          <DisplayData />
        </Await>
      </Suspense>
    </div>
  );
}
```

**Comment by DatTN95** [comment]
sorry but i have a problem, i using tanstack table, when i removed item in page number is 7 and call revalidator.revalidate();
page number back to 1, i need keep page 7

**Comment by renatobenks** [comment]
@DatTN95 Ideally, you should be put your search params in the URL through a query string, so you would be good to go!
